nilearn/tests/test_niimg_conversions.py

```diff
@@ -171,15 +190,15 @@ def test_concat_niimgs():
 
     # Regression test for #601. Dimensionality of first image was not checked
     # properly
-    assert_raises_regex(TypeError, 'Data must be a 3D Niimg-like object but '
-                        'you provided an image of shape',
+    assert_raises_regex(DimensionError, 'Data must be a 4D Niimg-like object but '
+                        'you provided',
                         _utils.concat_niimgs, [img4d], ensure_ndim=4)
 
     # check basic concatenation with equal shape/affine
     concatenated = _utils.concat_niimgs((img1, img3, img1))
 
-    assert_raises_regex(TypeError, 'Data must be a 3D Niimg-like object but '
-                        'you provided an image of shape',
+    assert_raises_regex(DimensionError, 'Data must be a 4D Niimg-like object but '
```

Comment:
Minor comment: looks like the message is slightly misleading for non-homogeneous lists:
In this case it will say "you provided a list of 4D images" although the first image is 3D...